jeremiahsnapp/chef-load
# Adding functionality for random data

# How to build and test:
In chef-load directory:
`hab package build .`
Unless there's a compile error, this builds a binary of the package you're working on. Find the hart file this command produced in the `results` folder, and move it to your a2 folder, with a command like the following
`cp results/elizabethuselton-chef-load-4.0.0-20171212193658-x86_64-linux.hart /Users/elizabethuselton/code/go/src/github.com/chef/a2/results/.`

Then, in an a2 studio:
`hab pkg install results/elizabethuselton-chef-load-4.0.0-20171212193658-x86_64-linux.hart  -b` but with whatever the hart file's name is.
Create a config file with `chef-load --sample-config > chef-load.conf`
After that, run `chef_load_nodes 10` to create 10 nodes.
`curl $ELASTICSEARCH_URL/node-state/_count` to see your number of nodes,
and `curl $ELASTICSEARCH_URL/node-state/_search` to see them.

# Open questions

1. Why is `organization_name` `demo` even with the randomized data? 
1. Why is `platform` `redhat` even with randomized data? 
Tentative answer: It seems it's pulling this from the `converge-success.json` and failing to override them, and `environment` is not specified in that file, meaning it doesn't get written over.
1. Has it always been this way? Possibly org and platform name have always been coming from `converge-success.json` even though they are supposed to be overwritten in the non-random data path by the same means that we are overwriting them in the random data path. https://github.com/chef/chef-load/blob/master/chef-client-run.go#L47
1. We know that chef_load is using `go-chef/chef`'s Node Struct. We are only seeing `Environment` get updated with our chef-load information, and we are wondering if this is because `Environment` is a top-level attribute in the Node Struct. None of the other fields are, though platform is allegedly in the ohai data which is getting saved as `AutomaticAttributes` but we can't figure out what exactly `AutomaticAttributes` is, where it comes from, or what it is composed of. 

Where We Left It:
1. Environment is successfully randomizing.
2. Organization and Platform are being randomized and integrated into the node data the same way they were in the non-random path, but they are not working. This is the current struggle.
3. Roles has code written for randomizing, but not hooked up, because we haven't found where in the node structure it should be. It's commented out so things compile.
4. No random seed has been added yet.
